cloudify_cli/commands/upgrade.py

```diff
@@ -88,26 +73,96 @@ def upgrade(validate_only,
                     task_retries=task_retries,
                     task_retry_interval=task_retry_interval,
                     task_thread_pool_size=task_thread_pool_size)
+        logger.info('Upgrade validation completed successfully')
 
     if not validate_only:
-        logger.info('Starting manager upgrade process...')
-        _put_workflow_state_file(is_upgrade=True)
-
-        logger.info('Executing manager upgrade...')
-        env.execute('install',
-                    task_retries=task_retries,
-                    task_retry_interval=task_retry_interval)
-
-    logger.info('Deactivating manager maintenance-mode...')
-    client.maintenance_mode.deactivate()
+        try:
+            logger.info('Executing Manager upgrade...')
+            env.execute('install',
+                        task_retries=task_retries,
+                        task_retry_interval=task_retry_interval,
+                        task_thread_pool_size=task_thread_pool_size)
+        except Exception as e:
+            msg = 'Failed upgrading Manager. Error: {0}'.format(e)
+            error = exceptions.CloudifyCliError(msg)
+            error.possible_solutions = [
+                "Rerun Manager upgrade command 'cfy upgrade'",
+                "Execute rollback command 'cfy rollback'"
+            ]
+            raise error
+
+        fabric_env = {
+            "host_string": management_ip,
+            "user": inputs['ssh_user'],
+            "key_filename": inputs['ssh_key_filename']
+        }
+        manager_node = next(node for node in env.storage.get_nodes()
+                            if node.id == 'manager_configuration')
+        bs.upload_resources(manager_node, fabric_env, inputs['public_ip'],
```

Comment:
Please note that this fires requests against the REST service, so perhaps maintenance mode should be deactivated first.